src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -240,149 +240,187 @@ private static ImmutableArray<TextChangeRange> Merge(IReadOnlyList<ImmutableArra
+            else if (newChanges.IsEmpty)
+            {
+                return oldChanges;
+            }
 
-            int oldIndex = 0;
-            int newIndex = 0;
-            int oldDelta = 0;
+            var builder = ArrayBuilder<TextChangeRange>.GetInstance();
 
-nextNewChange:
-            if (newIndex < newChanges.Length)
-            {
-                var newChange = newChanges[newIndex];
+            var oldChange = oldChanges[0];
+            var newChange = newChanges[0];
 
-nextOldChange:
-                if (oldIndex < oldChanges.Length)
-                {
-                    var oldChange = oldChanges[oldIndex];
+            var oldIndex = 0;
+            var newIndex = 0;
 
-tryAgain:
-                    if (oldChange.Span.Length == 0 && oldChange.NewLength == 0)
-                    {
-                        // old change is a non-change, just ignore it and move on
-                        oldIndex++;
-                        goto nextOldChange;
-                    }
-                    else if (newChange.Span.Length == 0 && newChange.NewLength == 0)
-                    {
-                        // new change is a non-change, just ignore it and move on
-                        newIndex++;
-                        goto nextNewChange;
-                    }
-                    else if (newChange.Span.End < (oldChange.Span.Start + oldDelta))
+            // The sum of characters inserted by old changes minus characters deleted by old changes.
+            // This value must be adjusted whenever characters from an old change are added to `builder`.
+            var oldDelta = 0;
+
+            // In this loop we "zip" together potentially overlapping old and new changes.
+            // It's important that when overlapping changes are found, we don't consume past the end of the overlapping section until the next iteration.
+            // so that we don't miss scenarios where the section after the overlap we found itself overlaps with another change
+            // e.g.:
+            // [ ------- oldChange1 ------ ]
+            // [ -- newChange1 -- ]   [ -- newChange2 --]
+            while (true)
+            {
+                if (oldChange.Span.Length == 0 && oldChange.NewLength == 0)
+                {
+                    if (tryGetNext(oldChanges, ref oldIndex, out oldChange)) continue;
+                    else break;
+                }
+                else if (newChange.Span.Length == 0 && newChange.NewLength == 0)
+                {
+                    if (tryGetNext(newChanges, ref newIndex, out newChange)) continue;
+                    else break;
+                }
+                else if (newChange.Span.End < oldChange.Span.Start + oldDelta)
+                {
+                    // new change is entirely before old change
+                    addAdjustedNewChange(newChange);
+                    if (tryGetNext(newChanges, ref newIndex, out newChange)) continue;
+                    else break;
```

Comment:
pictures for both before/after would be extremely helpful.  i.e. the comment explains the 'if' check above.  it does not explain teh code below.  tbh, while i can work my brain hard to convince myself this is right, just having a brief (visual) exlpanation woudl be super helpful.

Reply:
See https://github.com/dotnet/roslyn/blob/b63bda90e030af60b98a16da429d3fe048be3083/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/EditDistance.cs#L233-L476 for examples of how this can help.  Indeed, i'm fairly certain we had some text-span-merging code at one point, that even did this sort of thing (but i'm failing to find it now in our sprawling codebase :-/ ).